race-buffer/src/buffer.rs

```diff
@@ -0,0 +1,635 @@
+//! The writer of the RaceBuffer, which contains the actual buffer. Entries can be synchronously written and read/iterated
+//! from this struct.
+use crate::{get_seqn_index, num_missed, Entry, SeqNum, WholeEntry};
+//use core::iter::Iterator;
+use core::cmp::max;
+use core::fmt;
+use core::mem::size_of;
+use core::mem::MaybeUninit;
+use core::sync::atomic::fence;
+use core::sync::atomic::Ordering;
+
+/// Minimum allowed capacity of backing storage
+pub const MIN_STORAGE_CAP: usize = 4;
+
+/// Error for given storage size that is too small
+pub struct SizeError;
+
+impl fmt::Debug for SizeError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "Storage capacity must be at least {}", MIN_STORAGE_CAP,)
+    }
+}
+
+/// Round given length down to a power of 2
+#[inline]
+fn round_to_power_2(length: usize) -> usize {
+    let exp: usize = size_of::<usize>() * 8 - (length.leading_zeros() as usize) - 1;
+    1 << exp
+}
+
+#[derive(Debug)]
+#[repr(C)]
+/// Struct used to write to buffer
+pub struct RaceBuffer<'a, E>
+where
+    E: Entry,
+{
+    /// Sequence number of the next entry to be written
+    write_seqn: SeqNum,
+    /// Sequence number of the next entry to be overwritten
+    overwrite_seqn: SeqNum,
+    /// Backing storage
+    storage: &'a mut [MaybeUninit<E>],
+    /// Sequence number of next entry to be read from buffer
+    /// Note: when using RaceReader, this field is not used
+    read_seqn: SeqNum,
+    /// Indicates if backing storage should be truncated to a power of 2 length
+    /// in order to use optimized indexing
+    use_base_2_indexing: bool,
+}
+
+impl<'a, E> RaceBuffer<'a, E>
+where
+    E: Entry,
+{
+    /// Create new RaceBuffer. Returns error if storage size is not power of 2
```

Comment:
Is this documentation about the error outdated? It seems like the function is parameterized to accept a non-power-of-2 storage-length.